src/Workspaces/Core/MSBuild/MSBuild/Logging/MSBuildDiagnosticLogger.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using MSB = Microsoft.Build;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Logging
+{
+    internal class MSBuildDiagnosticLogger : MSB.Framework.ILogger
+    {
+        private string _projectFilePath;
+        private DiagnosticLog _log;
+        private MSB.Framework.IEventSource _eventSource;
+
+        public string Parameters { get; set; }
+        public MSB.Framework.LoggerVerbosity Verbosity { get; set; }
+
+        public void SetProjectAndLog(string projectFilePath, DiagnosticLog log)
+        {
+            _projectFilePath = projectFilePath;
+            _log = log;
+        }
+
+        private void OnErrorRaised(object sender, MSB.Framework.BuildErrorEventArgs e)
+        {
+            _log.Add(new MSBuildDiagnosticLogItem(WorkspaceDiagnosticKind.Failure, _projectFilePath, e.Message, e.File, e.LineNumber, e.ColumnNumber));
+        }
+
+        private void OnWarningRaised(object sender, MSB.Framework.BuildWarningEventArgs e)
+        {
+            _log.Add(new MSBuildDiagnosticLogItem(WorkspaceDiagnosticKind.Warning, _projectFilePath, e.Message, e.File, e.LineNumber, e.ColumnNumber));
+        }
+
+        public void Initialize(MSB.Framework.IEventSource eventSource)
+        {
+            _eventSource = eventSource;
```

Comment:
>_eventSource = eventSource; [](start = 12, length = 27)

Consider assert that _eventSource is null.

Reply:
Sure.

Reply:
FWIW, it appears that MSBuild *will* reuse this during batch build, so the assert was a good idea. We now clean up appropriately during `Shutdown()`.